contracts/RandomBeaconHistory.cdc

```diff
@@ -36,25 +49,167 @@ access(all) contract RandomBeaconHistory {
+        }
+
+        /// Finds the correct index to fill with the new random source. If a gap is detected, emits the
+        /// RandomHistoryMissing event.
+        ///
+        access(contract) view fun findGapAndReturnCorrectIndex(): UInt64 {
+
+            let currentBlockHeight = getCurrentBlock().height
+            // correct index to fill with the new random source
+            // so that eventually randomSourceHistory[correctIndex] = inputRandom
+            let correctIndex = currentBlockHeight - RandomBeaconHistory.lowestHeight!
+
+            // if a new gap is detected, emit an event
+            var arrayLength = UInt64(RandomBeaconHistory.randomSourceHistory.length)
+            if correctIndex > arrayLength {
+                let gapStartHeight = RandomBeaconHistory.lowestHeight! + arrayLength
+                emit RandomHistoryMissing(blockHeight: currentBlockHeight, gapStartHeight: gapStartHeight)
+            }
+            return correctIndex
+        }
+
+        /// Backfills possible empty entries (gaps) in the history array starting from the stored `gapStartIndex`,
+        /// using `randomSource` as a seed for all entries.
+        /// If there are no gaps, `gapStartIndex` is just updated to `RandomBeaconHistory`'s length.
+        //
+        /// When backfilling, all entries use the same entropy. Each entry is extracted from `randomSource` using
+        /// successive hashing. This makes sure the entries are all distinct although they provide
+        /// the same entropy.
+        //
+        /// gaps only occur in the rare event of a system transaction failure. In this case, entries are still
+        /// filled using a source not known at the time of block execution, which guarantees unpredicatability.
+        access(contract) fun backfill(randomSource: [UInt8]) {
+
+            let correctIndex = self.findGapAndReturnCorrectIndex()
+            var arrayLength = UInt64(RandomBeaconHistory.randomSourceHistory.length)
+            // optional optimization for the happy common path: if no gaps are detected,
+            // backfilling isn't needed, return early
+            if correctIndex == self.gapStartIndex  {
+                self.gapStartIndex = arrayLength + 1
+                return
+            }
+
+            // If a new gap is detected in the current transaction, fill the gap with empty entries.
+            // This happens in the rare case where a new gap occurs because of a system transaction failure.
+            while correctIndex > arrayLength {
+                RandomBeaconHistory.randomSourceHistory.append([])
+                arrayLength = arrayLength + 1
+            }
+
+            var newEntry = randomSource
+            var index = self.gapStartIndex
+            var count = 0 as UInt64
+            while count < self.maxEntriesPerCall {
+                // move to the next empty entry
+                while index < arrayLength && RandomBeaconHistory.randomSourceHistory[index] != [] {
```

Comment:
Here and elsewhere: Use a cheap `.length > 0` to check for non-emptiness, instead of creating a whole new empty array, each time:

```suggestion
                while index < arrayLength && RandomBeaconHistory.randomSourceHistory[index].length > 0 {
```